src/micro_monitoring.c

```diff
@@ -0,0 +1,342 @@
+    }
+
+    else if (nats_IsStringEmpty(name) && nats_IsStringEmpty(id))
+        return new_dotted_subject(newSubject, 2, MICRO_API_PREFIX, verb);
+    else if (nats_IsStringEmpty(id))
+        return new_dotted_subject(newSubject, 3, MICRO_API_PREFIX, verb, name);
+    else
+        return new_dotted_subject(newSubject, 4, MICRO_API_PREFIX, verb, name, id);
+}
+
+static microError *
+add_internal_handler(microService *m, const char *verb, const char *kind,
+                     const char *id, const char *name, microRequestHandler handler)
+{
+    microError *err = NULL;
+    char *subj = NULL;
+
+    err = micro_new_control_subject(&subj, verb, kind, id);
+    if (err != NULL)
+        return err;
+
+    microEndpointConfig cfg = {
+        .Subject = subj,
+        .Name = name,
+        .Handler = handler,
+    };
+    err = micro_add_endpoint(NULL, m, "", &cfg, true);
+    NATS_FREE(subj);
+    return err;
+}
+
+// __verbHandlers generates control handlers for a specific verb. Each request
+// generates 3 subscriptions, one for the general verb affecting all services
+// written with the framework, one that handles all services of a particular
+// kind, and finally a specific service instance.
+static microError *
+add_verb_handlers(microService *m, const char *verb, microRequestHandler handler)
+{
+    microError *err = NULL;
+    char name[1024];
+
+    snprintf(name, sizeof(name), "%s-all", verb);
+    err = add_internal_handler(m, verb, "", "", name, handler);
+    if (err == NULL)
+    {
+        snprintf(name, sizeof(name), "%s-kind", verb);
+        err = add_internal_handler(m, verb, m->cfg->Name, "", name, handler);
+    }
+    if (err == NULL)
+    {
+        err = add_internal_handler(m, verb, m->cfg->Name, m->id, verb, handler);
+    }
+    return err;
+}
+
+// name and sep must be a string literal
+#define IFOK_attr(_name, _value, _sep)                                  \
+    IFOK(s, natsBuf_Append(buf, "\"" _name "\":\"", -1));               \
+    IFOK(s, natsBuf_Append(buf, (_value) != NULL ? (_value) : "", -1)); \
+    IFOK(s, natsBuf_Append(buf, "\"" _sep, -1));
+
+static microError *
+marshal_ping(natsBuffer **new_buf, microService *m)
+{
+    natsBuffer *buf = NULL;
+    natsStatus s;
+
+    s = natsBuf_Create(&buf, 1024);
+    if (s == NATS_OK)
+    {
+        s = natsBuf_Append(buf, "{", -1);
+        IFOK_attr("name", m->cfg->Name, ",");
+        IFOK_attr("version", m->cfg->Version, ",");
+        IFOK_attr("id", m->id, ",");
+        IFOK_attr("type", MICRO_PING_RESPONSE_TYPE, "");
+        IFOK(s, natsBuf_AppendByte(buf, '}'));
+    }
+
+    if (s != NATS_OK)
+    {
+        natsBuf_Destroy(buf);
+        return micro_ErrorFromStatus(s);
+    }
+
+    *new_buf = buf;
+    return NULL;
+}
+
+static microError *
+marshal_info(natsBuffer **new_buf, microServiceInfo *info)
+{
+    natsBuffer *buf = NULL;
+    natsStatus s;
+
+    s = natsBuf_Create(&buf, 4096);
+    IFOK(s, natsBuf_Append(buf, "{", -1));
```

Comment:
When there is a single character, I often use natsBuf_AppendByte() instead, saves from internally calling strlen, or alternatively, pass 1 instead of -1.